File: test/key_test.exs
defmodule Karabinex.KeyTest do
  use ExUnit.Case

  alias Karabinex.Key

  describe "key definition parsing" do
    test "parses simple keys" do
      assert Key.parse(%Key{}, "x") == %Key{code: {:regular, "x"}}
    end

    test "modifiers: option" do
      assert Key.parse(%Key{}, "M-i") == %Key{
               code: {:regular, "i"},
               modifiers: MapSet.new([:option])
             }

      assert Key.parse(%Key{}, "⌥-i") == %Key{
               code: {:regular, "i"},
               modifiers: MapSet.new([:option])
             }
    end

    test "modifiers: command" do
      assert Key.parse(%Key{}, "⌘-i") == %Key{
               code: {:regular, "i"},
               modifiers: MapSet.new([:command])
             }
    end

    test "modifiers: control" do
      assert Key.parse(%Key{}, "C-q") == %Key{
               code: {:regular, "q"},
               modifiers: MapSet.new([:control])
             }

      assert Key.parse(%Key{}, "^-q") == %Key{
               code: {:regular, "q"},
               modifiers: MapSet.new([:control])
             }
    end

    test "modifiers: shift" do
      assert Key.parse(%Key{}, "S-c") == %Key{
               code: {:regular, "c"},
               modifiers: MapSet.new([:shift])
             }
    end

    test "modifiers: meh" do
      %Key{code: {:regular, "o"}, modifiers: modifiers} = Key.parse(%Key{}, "Meh-o")

      assert MapSet.equal?(MapSet.new([:shift, :option, :control]), modifiers)
    end

    test "modifiers: hyper" do
      %Key{code: {:regular, "b"}, modifiers: modifiers} = Key.parse(%Key{}, "H-b")

      assert Key.parse(%Key{}, "H-b") == Key.parse(%Key{}, "M-C-S-⌘-b")
      assert Key.parse(%Key{}, "H-b") == Key.parse(%Key{}, "✦-b")

      assert MapSet.equal?(MapSet.new([:shift, :option, :control, :command]), modifiers)
    end
  end
end

File: tests/key.test.ts
import { describe, expect, test } from "bun:test";
import { Key } from "../../src/key";
import { makeKeyCodes } from "../testUtils";

describe("key", () => {
  test("parses modifiers and code", () => {
    const codes = makeKeyCodes();
    const key = Key.parse("C-S-a", codes);
    expect(key.code.code).toBe("a");
    expect(key.modifiers).toEqual(["control", "shift"]);
  });

  test("parses Meh alias", () => {
    const codes = makeKeyCodes();
    const key = Key.parse("Meh-a", codes);
    expect(key.modifiers).toEqual(["option", "control", "shift"]);
    expect(key.readableName()).toBe("meh-a");
  });

  test("parses Hyper alias", () => {
    const codes = makeKeyCodes();
    const key = Key.parse("H-a", codes);
    expect(key.modifiers).toEqual(["command", "option", "control", "shift"]);
    expect(key.readableName()).toBe("hyper-a");
  });

  test("rejects duplicate modifiers", () => {
    const codes = makeKeyCodes();
    expect(() => Key.parse("C-C-a", codes)).toThrow();
  });

  test("codeSpec selects key type", () => {
    const codes = makeKeyCodes({
      regular: ["a"],
      consumer: ["volume_up"],
      pointer: ["button1"],
    });
    const regular = Key.parse("a", codes);
    const consumer = Key.parse("volume_up", codes);
    const pointer = Key.parse("button1", codes);

    expect(regular.codeSpec()).toEqual({ key_code: "a" });
    expect(consumer.codeSpec()).toEqual({ consumer_key_code: "volume_up" });
    expect(pointer.codeSpec()).toEqual({ pointing_button: "button1" });
  });

  test("rejects unknown key code", () => {
    const codes = makeKeyCodes();
    expect(() => Key.parse("unknown", codes)).toThrow();
  });
});
